### PR TITLE
Planner variations - simpler deco stop time calculation and variables use for variations

### LINKED ISSUE #878: Planner variations - list of issues
### Describe the issue:

- [x] Bug
- [x] Change request
- [x] New feature request
- [ ] Discussion request

### Issue long description:

@atdotde 
I'm going to open an issue to collect issues and feature requests in the current planner variations implementation. Just to make one thing clear: I'm doing this because I like the feature a lot and would like to improve it :-)
Here is what I have up to now:
1. Plan update issues described here #874 
2. solved
3. Possible mistake in calculating variations:
Additional deco stops are not considered.
If I look at this code here:
https://github.com/Subsurface-divelog/subsurface/blob/master/qt-models/diveplannermodel.cpp#L981
I would say you are ignoring possible additional stops in mid and max. This leads to underestimated delta time.
4. Calculating average of +- results but displaying it as + results.
My understanding is that today you calculate results for +-1m and +-1min, then you average these separately but then you print it as "+ x.xx /m + x.xx /min". Is this approach good at all? Maybe focus on +Xm / +Xmin only?
5. When printing result for +1min for the user it looks like this is a delta to total runtime. But in reality today it is just a delta for decostop time. I would prefer to print the real delta for runtime by simply adding the bottom time delta (today +-1min hard coded) as well.
6. Honestly now that I start to like the feature I want to be able to configure the delta depth/bottom time. Would be ok for me to only use one UI element (combobox) for both, linking it together like "+1m/+1min", "+2m/2min", "+3m/3min" or two separate UI elements.
7. Preparation for imperial units (ft) completely missing.

### PR DESCRIPTION
**Pull request by sfuchs79**
### Describe the pull request:

- [ ] Bug fix
- [x] Functional change
- [ ] New feature
- [x] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

A few change proposals for the planner variations to be discussed with @atdotde 

BTW:
I really appreciate this added string "Stop times" in the planner notes. Now this is ok for me because it is clear that we are NOT printing runtime variations.

### Related issues:

Closes #878

### CODE REVIEW

**Review by atdotde — APPROVED** [review]
Ah, no. I intended to approve the first commit, not the whole pull request.

I like the first one, this is indeed much simpler.

I don't see the point of the second except for a fundamental opposition to use goto. In my understanding, this is exactly the use case that goto in C was intended for (break out in sort of error condition but do necessary cleanup).

I am strongly against showing the variation for 3ft in imperial units: The diver is not supposed to actually change the depth by 1m or 1 respectively 3ft. She is supposed to do it by any (small enough) amount and then do the multiplication in her head. But that becomes more complicated when you list the result for 3 units rather than one unit. I am agnostic to actually changing the depth in the plan or doing the multiplication only in the printout. The difference in numbers will be much smaller than the precision of this kind of calculation anyway (even when computing it for 1ft where there is a factor of 3 between m and ft).

Four is fine again.

So: I would cherry pick 1 and 4 and could be convinced to like a version of three if it computes for 1 rather than three feet.

**Comment by sfuchs79** [comment]
Thanks :-)
Please wait with merging. Found s.th. else.

**Comment by sfuchs79** [comment]
Ok, now ready.
I now also put the "config" and the m/ft decision thing into one single place only. So if s.o. want's to change this, it's very low effort.

**Comment by sfuchs79** [comment]
Ah, ok, ready you other comment. Sure I'm open for any discussion here. Give me a few more minutes. I will try to push s.th. which could be the right compromise.

**Comment by sfuchs79** [comment]
Ok, set it to 1ft again. This is now much simpler (see comment above) :-)
I finally don't care about this "xft" discussion. I think finally the "imperial" users should decide.
You really want the "goto" back? I thought "goto" is only there in C to please ex-BASIC users? ;-)

**Comment by sfuchs79** [comment]
Yeah, ok, you're right. There are indeed a couple of "goto" use cases in our code...
I'll revert the change.

**Comment by sfuchs79** [comment]
Ok, done.
This could be now a version we both like. Unless I made any other stupid mistake.

We should also inform @dirkhh that already the PR before from Robert created a new string and kindly ask him to recreate the source strings :-)

File: qt-models/diveplannermodel.cpp [hunk 1]
```diff
@@ -986,32 +986,29 @@ struct divedatapoint * DivePlannerPointsModel::cloneDiveplan(struct diveplan *pl
...
+		midsum += mid->time;
 		++mid;
+	}
+	while (max->depth) {
+		maxsum += max->time;
 		++max;
 	}
-#ifdef SHOWSTOPVARIATIONS
+
+	leftsum = midsum - minsum;
```
Comment (atdotde) [inline qt-models/diveplannermodel.cpp:1008]: Now, I remember, why I had the more complicated logic: At some point, I was printing the differences for all individual stops. But since we ended up only using the aggregated values, this is no longer needed.

File: qt-models/diveplannermodel.cpp [hunk 2]
```diff
@@ -984,7 +984,7 @@ struct divedatapoint * DivePlannerPointsModel::cloneDiveplan(struct diveplan *pl
 	return last_segment;
 }
 
-int DivePlannerPointsModel::analyzeVariations(struct decostop *min, struct decostop *mid, struct decostop *max, const char *unit)
+int DivePlannerPointsModel::analyzeVariations(struct decostop *min, struct decostop *mid, struct decostop *max, QString unit)
```
Comment (atdotde) [inline qt-models/diveplannermodel.cpp]: Why this change? What is the advantage to go through the "fatter" class QString instead of *char?

File: qt-models/diveplannermodel.cpp [hunk 3]
```diff
@@ -1034,60 +1034,66 @@ void DivePlannerPointsModel::computeVariations(struct diveplan *original_plan, s
...
 		int my_instance = ++instanceCounter;
 		cache_deco_state(&ds, &save);
+		
+		int delta_time = 60;
+		int delta_depth;
+		QString depth_units;
+
+		if (prefs.units.length == units::METERS) {
+			delta_depth = 1000; // 1m
+			depth_units = QString("m");
```
Comment (atdotde) [inline qt-models/diveplannermodel.cpp]: Now, that we have this, shouldn't this be marked for translation?

File: qt-models/diveplannermodel.cpp [hunk 4]
```diff
@@ -1034,60 +1034,66 @@ void DivePlannerPointsModel::computeVariations(struct diveplan *original_plan, s
...
-		if (prefs.units.length == units::METERS)
-			sprintf(buf, ", %s + %d:%02d /m + %d:%02d /min", deco_time, FRACTION(analyzeVariations(shallower, original, deeper, "m"),60),
-				FRACTION(analyzeVariations(shorter, original, longer, "min"), 60));
-		else
-			sprintf(buf, ", %s + %d:%02d /ft + %d:%02d /min", deco_time,
-				FRACTION(analyzeVariations(shallower, original, deeper, "ft") * feet_to_mm(1.0) / 1000,60),
-				FRACTION(analyzeVariations(shorter, original, longer, "min"), 60));
+		sprintf(buf, ", %s: + %d:%02d /%s + %d:%02d /min", tr("Stop times").toUtf8().data(),
+			FRACTION(analyzeVariations(shallower, original, deeper, depth_units), 60), depth_units.toUtf8().data(),
+			FRACTION(analyzeVariations(shorter, original, longer, QString("min")), 60));
```
Comment (atdotde) [inline qt-models/diveplannermodel.cpp]: tr() here as well.

**Review by atdotde — COMMENTED** [review]
Just two small things i missed this morning.

File: qt-models/diveplannermodel.cpp (hunk 2 above)
Comment (bstoeger) [reply qt-models/diveplannermodel.cpp]: By the way: is there any agreement on whether to pass by const reference (`const QString &unit`) or by value (`QString unit`)? It doesn't matter too much, since `QString`s are COW/reference counted (an unfortunate design choice in my opinion). But if you can avoid it - why do reference counting?

**Comment by dirkhh** [comment]
On the subject of goto...
That's a very useful construct if used the right way.
You always jump forward, usually to end processing in this function but do some shared cleanup

**Comment by sfuchs79** [comment]
@dirkhh 
I apologize for insulting the C "goto" statement ;-)
Pushed another update introducing the tr(), introducing usage of duration_t and depth_t unit and going back to const char* for one functions argument.
@atdotde please review again.

**Comment by atdotde** [comment]
Beautiful! Even the orchestra is beautiful!
